vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go

```diff
@@ -174,6 +173,9 @@ func (d *Deleter) deletePV(pv *v1.PersistentVolume) error {
 				return fmt.Errorf("Error deleting PV %q: %v", pv.Name, err.Error())
 			}
 		}
+		// Now we're certain the PV was deleted in API and can update cache
+		d.Cache.CleanPV(pv)
```

Comment:
I worry there may be a race in either case.
1) clean the cache entry before `d.APIUtil.DeletePV`: if the API call fails, then it won't be retried.
2) clean the cache entry after `d.APIUtil.DeletePV`: could we end up with the [original problem](https://bugzilla.redhat.com/show_bug.cgi?id=2032924) where [DeletePVs is called](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/diskmaker/controllers/deleter/reconcile.go#L71) at the wrong time and end up starting a second cleanup job before the PV is removed from the cache?

There is also a problem with this ordering, in that [DeletePV](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/cache/cache.go#L78) removes the 'cleaned' cache entry, [CleanPV](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/cache/cache.go#L99) adds it back as true, which means the next time LSO attempts to clean a PV with this name, it [will think it was already cleaned up successfully](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/cache/cache.go#L109) and do nothing. So it causes us not to clean the PV the next time.

Is it possible to leave the call to `CleanPV` before the call to `DeletePV`, and consider the cleanup part successful, but still retry the call to `DeletePV` if the API call fails? This doesn't seem so simple, because in order to [check the CSSucceeded state](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L166) we actually [remove the cleanup job status](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L160) when checking it!
There is also the option to do a real refactoring (see [bug 2095077](https://bugzilla.redhat.com/show_bug.cgi?id=2095077)) but that might not be a good backport candidate. Still, any other option I see leads to more complexity and potential gaps.

Reply:
> 2. clean the cache entry after `d.APIUtil.DeletePV`: could we end up with the [original problem](https://bugzilla.redhat.com/show_bug.cgi?id=2032924) where [DeletePVs is called](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/diskmaker/controllers/deleter/reconcile.go#L71) at the wrong time and end up starting a second cleanup job before the PV is removed from the cache?

I think this is not really a problem, it should not be possible to trigger `DeletePVs` more than once at a time from the `Reconcile` loop. But the ordering problem I mentioned does look like it will cause issues, at least for the current implementation.

Reply:
> Is it possible to leave the call to `CleanPV` before the call to `DeletePV`, and consider the cleanup part successful, but still retry the call to `DeletePV` if the API call fails? This doesn't seem so simple, because in order to [check the CSSucceeded state](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L166) we actually [remove the cleanup job status](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L160) when checking it! 

Changing something while *just checking* is strange indeed. But the cleanup job status is [added back](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L211) if it's not a success. So I think there are two options:

1) Add a function similar to [RemoveStatus()](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L388) which would just report the status and not delete it making the code more readable. But that will require some refactoring.

2) Leave the cache cleaning *before* API call but [on failure](https://github.com/openshift/local-storage-operator/blob/664054169e9d761e2d109d4c5f9bf4d7a1c4397c/vendor/sigs.k8s.io/sig-storage-local-static-provisioner/pkg/deleter/deleter.go#L172) add it back along with the cleanup job. 


> There is also the option to do a real refactoring (see [bug 2095077](https://bugzilla.redhat.com/show_bug.cgi?id=2095077)) but that might not be a good backport candidate. Still, any other option I see leads to more complexity and potential gaps.

That's the best but most time consuming approach. Looking at your suggestions in BZ I think option 2 (refactor to use events) would be the best. Just because option 1 means more code to maintain and option 3 reminds me of PV controllers resync approach which turned out to be "not the best".